server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java

```diff
@@ -119,61 +134,114 @@ public void exit() {
-      String gradleVersion = "";
-      GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
-      Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
-      LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
-      if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
-        gradleVersion = preferences.getGradleVersion();
-      } else {
-        gradleVersion = connector.getGradleVersion(rootUri);
-      }
+    GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
+    Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
+    LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
+
+    String gradleVersion;
+    if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
+      gradleVersion = preferences.getGradleVersion();
+    } else {
+      gradleVersion = connector.getGradleVersion(rootUri);
+    }
+
+    if (StringUtils.isEmpty(gradleVersion)) {
+      return null;
+    }
+
+    map = TelemetryUtils.getMetadataMap("gradleVersion", gradleVersion);
+    LOGGER.log(Level.INFO, "Gradle version: " + gradleVersion, map);
+
+    String latestCompatibleVersion = Utils.getLatestCompatibleJavaVersion(gradleVersion);
+    String oldestCompatibleVersion = Utils.getOldestCompatibleJavaVersion();
+
+    if (StringUtils.isNotBlank(latestCompatibleVersion)) {
 
-      if (StringUtils.isNotBlank(gradleVersion)) {
-        map = TelemetryUtils.getMetadataMap("gradleVersion", gradleVersion);
-        LOGGER.log(Level.INFO, "Gradle version: " + gradleVersion, map);
-        String highestJavaVersion = Utils.getHighestCompatibleJavaVersion(gradleVersion);
-        File jdkInstallation = getJdkToLaunchDaemon(preferences.getJdks(), highestJavaVersion);
-        if (jdkInstallation != null) {
-          preferences.setGradleJavaHome(jdkInstallation.getAbsolutePath());
+      // Use GradleJavaHome if compatible
+      if (gradleJavaHome != null) {
+
+        try {
+          String gradleJavaHomeVersion = JavaUtils.getJavaVersionFromFile(gradleJavaHome);
+          if (
+              JavaUtils.isCompatible(
+                  gradleJavaHomeVersion,
+                  oldestCompatibleVersion,
+                  latestCompatibleVersion
+              )
+          ) {
+            return gradleJavaHome;
+          }
+        } catch (IOException e) {
+          LOGGER.severe("Invalid GradleJavaHome: " + e.getMessage());
         }
+
+      }
+
+      // Pick a compatible JDK from the JDKs available in Preferences
+      if (preferences.getJdks() != null && preferences.getJdks().isEmpty()) {
+        return getJdkToLaunchDaemon(
+            preferences.getJdks(),
+            oldestCompatibleVersion,
+            latestCompatibleVersion
+        );
       }
+
     }
+
+    // Notify client, for no compatible JDK can be found
+    new ClientNotifier(client).sendNotification(
```

Comment:
I'm thinking that if this is over design at this stage for the `ClientNotifier`? Does that mean we need to create a new notifier instance every time we want to send some msg to the client?

I like the idea of wrap things like taskid & originId into it. Maybe at current stage, just call the method from `client` directly? In the future, if we found that this is not enough, we can refactor to introduce sth like the notifier. WDYT?

Reply:
If there is no requirement for the `taskId` and `originId` at this stage, I could make the `sendNotification` method static and there won't be the need to create new instances of ClientNotifier.

Reply:
I would prefer directly using `client`.

Reply:
Lifecycle service directly uses the client to send notifications with [847c8a2](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/847c8a25e0d409e03f2628beea5aa87f27c8dbd2)